labelbox/schema/data_row.py

```diff
@@ -17,6 +19,35 @@
 logger = logging.getLogger(__name__)
 
 
+class KeyType(str, Enum):
+    ID = 'ID'
+    """An existing CUID"""
+    GKEY = 'GKEY'
+    """A Global key, could be existing or non-existing"""
+    AUTO = 'AUTO'
+    """The key will be auto-generated. Only usable for creates"""
+
+
+class ResolvableId(BaseModel):
+    """
+    The ResolvableId class is a unique ID abstraction that allows us to reference
+    a DataRow by either a Global key or CUID
+    """
+    type: KeyType = KeyType.GKEY
+    value: str
+
+
+class DataRowSpec(BaseModel):
+    row_data: Union[str, dict]
+    external_id: Optional[str]
+    global_key: Optional[str]
+
+
+class DataRowUpsertItem(BaseModel):
+    id: ResolvableId
```

Comment:
We might consider abstracting this concept from the user and map the user-facing entity to this type which the backend expects.

One example is the user can optionally provide a key; otherwise we assume it will be auto-generated (on create).